modules/text2vec-databricks/vectorizer/objects.go

```diff
@@ -0,0 +1,83 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package vectorizer
+
+import (
+	"context"
+	"time"
+
+	"github.com/weaviate/tiktoken-go"
+	"github.com/weaviate/weaviate/entities/models"
+	"github.com/weaviate/weaviate/entities/moduletools"
+	"github.com/weaviate/weaviate/modules/text2vec-databricks/clients"
+	"github.com/weaviate/weaviate/modules/text2vec-databricks/ent"
+	"github.com/weaviate/weaviate/usecases/modulecomponents/text2vecbase"
+	objectsvectorizer "github.com/weaviate/weaviate/usecases/modulecomponents/vectorizer"
+
+	"github.com/weaviate/weaviate/usecases/modulecomponents/batch"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	MaxObjectsPerBatch = 2000 // https://platform.openai.com/docs/api-reference/embeddings/create
+	// time per token goes down up to a certain batch size and then flattens - however the times vary a lot so we
+	// don't want to get too close to the maximum of 50s
+	OpenAIMaxTimePerBatch = float64(10)
+)
+
+func New(client text2vecbase.BatchClient, logger logrus.FieldLogger) *text2vecbase.BatchVectorizer {
+	batchTokenizer := func(ctx context.Context, objects []*models.Object, skipObject []bool, cfg moduletools.ClassConfig, objectVectorizer *objectsvectorizer.ObjectVectorizer) ([]string, []int, bool, error) {
+		texts := make([]string, len(objects))
+		tokenCounts := make([]int, len(objects))
+		icheck := ent.NewClassSettings(cfg)
+
+		// TODO: Check how Databricks embeddings model tokenize the text
+		tke, err := tiktoken.EncodingForModel(icheck.Model())
+		if err != nil { // fail all objects as they all have the same model
+			tke, _ = tiktoken.EncodingForModel("text-embedding-ada-002")
+		}
+
+		// prepare input for vectorizer, and send it to the queue. Prepare here to avoid work in the queue-worker
+		skipAll := true
+		for i := range texts {
+			if skipObject[i] {
+				continue
+			}
+			skipAll = false
+			text := objectVectorizer.Texts(ctx, objects[i], icheck)
+			texts[i] = text
+			tokenCounts[i] = clients.GetTokensCount(icheck.Model(), text, tke)
```

Comment:
@antas-marcin do you why it is necessary for the batch tokenizer to return a count fo the input tokens too? Asking because the embeddings models in databricks uses wordpiece tokenization and I couldn't find an "official" (i.e like tiktoken) golang library for this

Reply:
>  Asking because the embeddings models in databricks uses wordpiece tokenization and I couldn't find an "official" (i.e like tiktoken) golang library for this

Some providers have a token limit per batch and minute - we somehow need the token count for the input to keep track how close we are to these limits.

If they have one, please ask them what a good approximation using toktoken would be - eg the toktoken count *1.3 as an approximation. This only has an effect on how many objects are put into one batch (batch token limit) or per minute. So it doesn't hurt if it is a crude approximation

Reply:
@dirkkul Databricks do not impose such a limit on their endpoint. How should we adapt this code then?

Reply:
you can set it to 0, like in cohere: https://github.com/weaviate/weaviate/blob/07762cca4f42ec5ac6f1c224ce28aa22c350bee1/modules/text2vec-cohere/vectorizer/objects.go#L49